### PR TITLE
fix #249541: PluginCreator improvements.

### PR DESCRIPTION
**Pull request by rjmatthews62**
Improved Help - added a very basic guide and external links.
Very simple autoindent implemented in editor - new lines line up with previous lines
"New" includes placeholders for description and version.
Refresh button on PluginManager forces modifications to be picked up
Make plugins stay on top while using Run

Note: incorporates #249631 (PluginCreator SaveAs)

### CODE REVIEW

File: manual/genManual.cpp [hunk 1]
```diff
@@ -319,8 +319,14 @@ static void writeOutput()
...
       QString out;
       addHeader(out);
-      out += "<h2>Score Elements</h2>\n"
-             "<ul>\n";
+      out += "<h2>Score Elements</h2>\n";
+      out += "<h3>Quick Guide</h3>\n";
+      out += "<p>Below are all the various classes you can use."
+             "<br>The main class is <a href='musescore.html'>MuseScore</a>."
+             "<br>Use 'New' to create a skeleton plugin."
+             "<br>Plugins are coded in <a href='http://doc.qt.io/qt-5/qmlapplications.html#what-is-qml'>QML</a>"
```
Comment (Jojo-Schmitz) [inline manual/genManual.cpp]: Not sure, do we want these strings to be translatable?

File: mscore/pluginManager.cpp [hunk 2]
```diff
@@ -67,9 +91,7 @@ void PluginManager::init()
             pluginList->setCurrentRow(0);
             pluginListItemChanged(pluginList->item(0), 0);
             }
-      connect(pluginList, SIGNAL(itemChanged(QListWidgetItem*)), SLOT(pluginLoadToggled(QListWidgetItem*)));
-      connect(pluginList, SIGNAL(currentItemChanged(QListWidgetItem*, QListWidgetItem*)),
-         SLOT(pluginListItemChanged(QListWidgetItem*, QListWidgetItem*)));
+
```
Comment (Jojo-Schmitz) [inline mscore/pluginManager.cpp]: can you remove that empty line?

File: mscore/pluginManager.cpp [hunk 3]
```diff
@@ -49,14 +50,37 @@ void PluginManager::init()
...
+//---------------------------------------------------------
+//   loadList - populate the listbox.
+//---------------------------------------------------------
+
+void PluginManager::loadList(bool forceRefresh)
+      {
+      QStringList saveLoaded; // If forcing a refresh, the load flags are lost. Keep a copy and reapply.
+      int n=preferences.pluginList.size();
+      if (forceRefresh && n>0) {
+            for(int i=0; i<n; i++) {
```
Comment (Jojo-Schmitz) [inline mscore/pluginManager.cpp]: space between `for`and `(`, spaces before and after `=``and `>`

File: manual/genManual.cpp (hunk 1 above)
Comment (rjmatthews62) [reply manual/genManual.cpp]: Well, it wasn't translatable before...

File: mscore/pluginManager.cpp (hunk 3 above)
Comment (rjmatthews62) [reply mscore/pluginManager.cpp]: Has anyone told you you're picky?

File: manual/genManual.cpp (hunk 1 above)
Comment (rjmatthews62) [reply manual/genManual.cpp]: Do you just want me to surround it with a "tr"?

File: mscore/pluginCreator.cpp [hunk 4]
```diff
@@ -306,6 +313,11 @@ void PluginCreator::runClicked()
       run->setEnabled(false);
 
       item = qobject_cast<QmlPlugin*>(obj);
+      msg("Plugin Details:\n");
+      msg("  Menupath: "+item->menuPath()+"\n");
+      msg("  Version: "+item->version()+"\n");
+      msg("  Description: "+item->description()+"\n");
+      if (item->requiresScore()) msg("  Requires Score\n");
```
Comment (Jojo-Schmitz) [inline mscore/pluginCreator.cpp]: same question reg. translatability

File: mscore/pluginManager.cpp [hunk 5]
```diff
@@ -49,14 +50,37 @@ void PluginManager::init()
...
+            }
+      preferences.updatePluginList(forceRefresh);
+      n = preferences.pluginList.size();
       pluginList->clear();
       for (int i = 0; i < n; ++i) {
             PluginDescription& d = preferences.pluginList[i];
             Shortcut* s = &d.shortcut;
             localShortcuts[s->key()] = new Shortcut(*s);
-
+            if (saveLoaded.contains(d.path)) d.load=true;
```
Comment (Jojo-Schmitz) [inline mscore/pluginManager.cpp]: better 2 lines, and spaces around the `=`

File: manual/genManual.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [reply manual/genManual.cpp]: As said, I'm not sure whether we want or need ot translatable. @lasconic?

File: mscore/pluginManager.cpp (hunk 3 above)
Comment (Jojo-Schmitz) [reply mscore/pluginManager.cpp]: I am by nature ;-)

File: manual/genManual.cpp (hunk 1 above)
Comment (rjmatthews62) [reply manual/genManual.cpp]: Yeah, that's in genmanual.cpp, and there are no "tr" features at all in there... I think that counts as a whole extra issue.

Comment (Jojo-Schmitz) [reply manual/genManual.cpp]: Yes, surely not your fault. But might make up yet another improvement. I'm just not sure whether it is worth the effort?

Comment (rjmatthews62) [reply manual/genManual.cpp]: Well, the issue is that that plugin help auto generates metadata in the header files... which are not translated.
So it's a whole new kettle of fish. Worth thinking about, but above and beyond the current context, I think,

Comment (Jojo-Schmitz) [reply manual/genManual.cpp]: Agreed

**Comment by rjmatthews62** [comment]
Posted update, rebased to keep to the one commit (I assume this is what we do?)
Did you know that the Travis-CI build is hanging in the 64 bit build job?
Waiting on "enter passphrase"

File: mscore/pluginManager.cpp (hunk 2 above)
Comment (rjmatthews62) [reply mscore/pluginManager.cpp]: I think I've fixed that one? Not sure where it thought the line was...

File: mscore/pluginManager.cpp (hunk 3 above)
Comment (Jojo-Schmitz) [reply mscore/pluginManager.cpp]: Ready for more pickiness? There are more spaces missing, around `+` and `=` later in your PR

File: mscore/pluginManager.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply mscore/pluginManager.cpp]: It is still in.

**Comment by Jojo-Schmitz** [comment]
Yes, that is epected (both, the rebase and one commit as well as Travis not building appimages for PRs

**Comment by rjmatthews62** [comment]
Is there anything outstanding on this PR that I've missed?

**Comment by Jojo-Schmitz** [comment]
Except for some pickyness reg. spaces before and after operators in the autoIndent() method as well as in runClicked() and load(), I see nothing.

**Comment by rjmatthews62** [comment]
How's that, then?

**Comment by Jojo-Schmitz** [comment]
Still a few left. Let me mark them...

File: mscore/pluginCreator.cpp [hunk 6]
```diff
@@ -306,10 +313,15 @@ void PluginCreator::runClicked()
       run->setEnabled(false);
 
       item = qobject_cast<QmlPlugin*>(obj);
+      msg(tr("Plugin Details:\n"));
+      msg(tr("  Menupath: ")+item->menuPath()+"\n");
+      msg(tr("  Version: ")+item->version()+"\n");
+      msg(tr("  Description: ")+item->description()+"\n");
```
Comment (Jojo-Schmitz) [inline mscore/pluginCreator.cpp]: this `+` and the ones in the lines before should get surrounded by spaces

File: mscore/pluginCreator.cpp [hunk 7]
```diff
@@ -430,19 +452,20 @@ void PluginCreator::load()
...
       {
-      if (created) {
+      if (saveas) {
             path = mscore->getPluginFilename(false);
             if (path.isEmpty())
                   return;
             }
       QFile f(path);
       QFileInfo fi(f);
+      msg(tr("Saving to:")+path+"\n");
```
Comment (Jojo-Schmitz) [inline mscore/pluginCreator.cpp]: same here

File: mscore/pluginManager.cpp [hunk 8]
```diff
@@ -48,15 +49,39 @@ void PluginManager::init()
...
-      preferences.updatePluginList();
+//---------------------------------------------------------
+//   loadList - populate the listbox.
+//---------------------------------------------------------
+
+void PluginManager::loadList(bool forceRefresh)
+      {
+      QStringList saveLoaded; // If forcing a refresh, the load flags are lost. Keep a copy and reapply.
       int n = preferences.pluginList.size();
+      if (forceRefresh && n>0) {
```
Comment (Jojo-Schmitz) [inline mscore/pluginManager.cpp]: the `>` should get surrounded by spaces

File: mscore/pluginManager.cpp [hunk 9]
```diff
@@ -48,15 +49,39 @@ void PluginManager::init()
...
+      n = preferences.pluginList.size();
       pluginList->clear();
       for (int i = 0; i < n; ++i) {
             PluginDescription& d = preferences.pluginList[i];
             Shortcut* s = &d.shortcut;
             localShortcuts[s->key()] = new Shortcut(*s);
-
+            if (saveLoaded.contains(d.path)) {
+                  d.load=true;
+                  }
```
Comment (Jojo-Schmitz) [inline mscore/pluginManager.cpp]: for a one-liner the braces are not needed

**Comment by Jojo-Schmitz** [comment]
And the same is true for #3291 I think

**Comment by Jojo-Schmitz** [comment]
Travis CI failure seems to be unrelated, seems @lasconic would need to check

**Comment by Jojo-Schmitz** [comment]
Now with #3289 merged this needs a rebase

**Comment by lasconic** [comment]
To rebase,

```
git checkout master
git fetch upstream
git merge upstream/master
git checkout fix_249541
git rebase master
// fix any conflict
git push -f origin fix_249541
```

**Comment by rjmatthews62** [comment]
Done.

**Comment by lasconic** [comment]
Cool thanks.

**Pull request merged**